File: api/routers/stats.py
from fastapi import APIRouter, Depends, Query, HTTPException
from models.schemas import StatsResponse, EvaluationSummary
from services.auth import verify_supabase_jwt
from services.database import get_supabase

router = APIRouter()


@router.get("/stats", response_model=StatsResponse)
async def stats(
    key_id: str = Query(..., description="表示するAPIキーのID"),
    offset: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    user: dict = Depends(verify_supabase_jwt),
):
    supabase = get_supabase()

    # key_id がログインユーザーのものか確認
    key_check = (
        supabase.table("api_keys")
        .select("id")
        .eq("id", key_id)
        .eq("user_id", user["id"])
        .limit(1)
        .execute()
    )
    if not key_check.data:
        raise HTTPException(status_code=404, detail="API key not found")

    # 件数
    count_result = (
        supabase.table("evaluations")
        .select("id", count="exact")
        .eq("api_key_id", key_id)
        .execute()
    )
    total = count_result.count or 0

    # 平均スコア
    avg_result = (
        supabase.table("evaluations")
        .select(
            "auto_score_relevance,"
            "auto_score_faithfulness,"
            "auto_score_completeness"
        )
        .eq("api_key_id", key_id)
        .not_.is_("auto_score_relevance", "null")
        .execute()
    )

    rows = avg_result.data or []
    avg_relevance = avg_faithfulness = avg_completeness = None
    if rows:
        avg_relevance = sum(r["auto_score_relevance"] for r in rows) / len(rows)
        avg_faithfulness = sum(r["auto_score_faithfulness"] for r in rows) / len(rows)
        avg_completeness = sum(r["auto_score_completeness"] for r in rows) / len(rows)

    # フィードバック率
    fb_result = (
        supabase.table("feedbacks")
        .select("rating, evaluations!inner(api_key_id)")
        .eq("evaluations.api_key_id", key_id)
        .execute()
    )
    feedbacks = fb_result.data or []
    positive_feedback_rate = None
    if feedbacks:
        positive = sum(1 for f in feedbacks if f["rating"] == 1)
        positive_feedback_rate = positive / len(feedbacks)

    # 最近の評価
    recent_result = (
        supabase.table("evaluations")
        .select(
            "id, question, answer, chunks, latency_ms,"
            "auto_score_relevance, auto_score_faithfulness, auto_score_completeness,"
            "created_at"
        )
        .eq("api_key_id", key_id)
        .order("created_at", desc=True)
        .range(offset, offset + limit - 1)
        .execute()
    )

    recent = [EvaluationSummary(**r) for r in (recent_result.data or [])]

    return StatsResponse(
        total=total,
        avg_relevance=avg_relevance,
        avg_faithfulness=avg_faithfulness,
        avg_completeness=avg_completeness,
        positive_feedback_rate=positive_feedback_rate,
        recent=recent,
    )

File: api-worker/src/routes/stats.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { HTTPException } from "hono/http-exception";
import type { AppVariables, Bindings } from "../types";
import { verifyJwt } from "../lib/auth";
import { StatsQuerySchema } from "../schemas";

type RecentRow = {
  id: string;
  question: string;
  answer: string;
  chunks: string; // JSON string
  latency_ms: number | null;
  auto_score_relevance: number | null;
  auto_score_faithfulness: number | null;
  auto_score_completeness: number | null;
  created_at: string;
};

export const statsRoute = new Hono<{
  Bindings: Bindings;
  Variables: AppVariables;
}>();

statsRoute.get(
  "/stats",
  verifyJwt,
  zValidator("query", StatsQuerySchema),
  async (c) => {
    const { key_id, offset, limit } = c.req.valid("query");
    const user = c.get("user");
    const db = c.env.DB;

    // key_id がログインユーザーのものか確認
    const keyCheck = await db
      .prepare(
        "SELECT id FROM api_keys WHERE id = ? AND user_id = ? LIMIT 1",
      )
      .bind(key_id, user.id)
      .first<{ id: string }>();

    if (!keyCheck) {
      throw new HTTPException(404, { message: "API key not found" });
    }

    // 件数 + 平均スコア (1クエリで集計)
    const agg = await db
      .prepare(
        `SELECT
           COUNT(*) AS total,
           AVG(auto_score_relevance) AS avg_relevance,
           AVG(auto_score_faithfulness) AS avg_faithfulness,
           AVG(auto_score_completeness) AS avg_completeness
         FROM evaluations
         WHERE api_key_id = ?`,
      )
      .bind(key_id)
      .first<{
        total: number;
        avg_relevance: number | null;
        avg_faithfulness: number | null;
        avg_completeness: number | null;
      }>();

    // フィードバック率
    const fb = await db
      .prepare(
        `SELECT
           COUNT(*) AS total,
           SUM(CASE WHEN f.rating = 1 THEN 1 ELSE 0 END) AS positive
         FROM feedbacks f
         JOIN evaluations e ON f.evaluation_id = e.id
         WHERE e.api_key_id = ?`,
      )
      .bind(key_id)
      .first<{ total: number; positive: number | null }>();

    const positive_feedback_rate =
      fb && fb.total > 0 ? (fb.positive ?? 0) / fb.total : null;

    // 最近の評価
    const recentResult = await db
      .prepare(
        `SELECT id, question, answer, chunks, latency_ms,
                auto_score_relevance, auto_score_faithfulness, auto_score_completeness,
                created_at
         FROM evaluations
         WHERE api_key_id = ?
         ORDER BY created_at DESC
         LIMIT ? OFFSET ?`,
      )
      .bind(key_id, limit, offset)
      .all<RecentRow>();

    const recent = (recentResult.results ?? []).map((r) => ({
      ...r,
      chunks: safeParseJson(r.chunks),
    }));

    return c.json({
      total: agg?.total ?? 0,
      avg_relevance: agg?.avg_relevance ?? null,
      avg_faithfulness: agg?.avg_faithfulness ?? null,
      avg_completeness: agg?.avg_completeness ?? null,
      positive_feedback_rate,
      recent,
    });
  },
);

function safeParseJson(text: string): unknown {
  try {
    return JSON.parse(text);
  } catch {
    return [];
  }
}
